Make float test more accurate, add integer performance tests.

Kuratius opened the pull request:
This doesn't contain an accuracy test for the integer square root functions, but I could add one if needed.

asiekierka commented on tests/system/hw_math/source/main.c:
```diff
@@ -7,6 +7,82 @@
 
+#define TEST_SIZE (1000)
+
+u32 dkp_sqrt32(int a)
```
Please avoid putting copy-pasted benchmarks for other projects in BlocksDS. Not only can they change the code at any time, which would make the copy paste outdated, it feels in bad taste.

This can be a separate project.

AntonioND replied:
Yeah, I agree with asie.

Kuratius replied:
I'll remove them then. I could put in a software implementation test instead and rename the output.

asiekierka replied:
Yeah, measuring against a simple C-based implementation sounds like a better idea and provides meaningful information to developers; it can also be used to validate correct outputs from the BlocksDS hardware accelerated implementation.

AntonioND commented on tests/system/hw_math/source/main.c:
```diff
@@ -49,6 +125,9 @@ int main(int argc, char **argv)
     printf("\n");
 
+    volatile float* vf=malloc(sizeof(float));
+    if (vf==0) printf("could not allocate volatile float!\n");
```
This isn't needed. All you need to do is put `volatile float vf;` as a global variable and write to it from below.

Kuratius replied:
I'll remove the Malloc call and perhaps also replace TIMER3_CR since it's also a pointer.

AntonioND replied:
Yeah, sorry, that's what I meant. You can use the global in all the loops.

AntonioND commented on tests/system/hw_math/source/main.c:
```diff
@@ -230,14 +417,271 @@ int main(int argc, char **argv)
+
+            sw_time += cpuEndTiming();
+
+            printf("BlocksDS sqrt64 time: %lu (%lu per op)\n", sw_time, sw_time / sw_iterations);
```
This won't fit in one line, you should split it into two so that numbers aren't cut in the middle.

AntonioND commented on tests/system/hw_math/source/main.c:
```diff
@@ -7,6 +7,187 @@
+    //which gives the source as  
+    //Square root by abacus algorithm, Martin Guy @ UKC, June 1985.
+    //From a book on programming abaci by Mr C. Woo.
+    //assert(("sqrt input should be non-negative", n > 0));
```
Don't leave this commented `assert()`. You should just make the function take unsigned ints as input and return unsigned ints.

AntonioND commented on tests/system/hw_math/source/main.c:
```diff
@@ -230,14 +417,271 @@ int main(int argc, char **argv)
+    }
+
+
+
```
Don't add too many spaces, indent blocks correctly.

AntonioND commented on tests/system/hw_math/source/main.c:
```diff
@@ -49,6 +235,8 @@ int main(int argc, char **argv)
     printf("qNaN: 0x%08lX\n", quiet_nan);
     printf("\n");
```
Don't add (or remove) empty lines for no reason, they just add noise to the git history.

AntonioND reviewed (commented):
You really need to try to match the code style of the repository when you contribute to projects.

In the case of this repository, don't leave many empty lines between functions (one is enough), try to match where braces are opened and closed, capitalise comments correctly... Take a look at other PRs and you will see that, for the most part, contributors imitate pre-existing code. I'm not going to complain about one brace here and there not matching other code, but this is too different, and I would just have to clean it up myself.

Kuratius commented:
I plan to work on this further, I wouldn't consider it finished. I assume to reduce noise/you getting unnecessary notifications it's probably not enough to mark it as a draft and request a review manually when necessary.  The option I have here is either to close the PR until it's done, or to split the changes, although then I'm not sure why the draft feature exists to begin with.

asiekierka commented:
The draft feature is fine; just keep us updated on the status. If you ever feel like the PR is not up to standard, feel free to close it.

Kuratius commented:
I've updated my tests and also added some tests for the inverse sqrt in https://github.com/blocksds/libnds/pull/156 . The file currently needs to be copied from there to run some of the tests.

I've modified the integer sqrt implementations for unsigned inputs and added an assembly routine for 32-bit integer sqrts. If you think the speed is good enough you may want the assembly routine as part of libnds, though. In this PR it is included as a separate file.  I believe at least for DSi users it outperforms the hardware sqrt unit, but I cannot currently test on DS. I'm also curious if choosing a better starting value with clz for it is possible, but I havent experimented with it, as it already seems to be ridiculously fast.

Kuratius commented:
I'll do some more experimentation with other methods and try to clean this up a bit more. The way the tests are selected could probably be done better than by choosing a new set of buttons every time (& the printed text could be clearer).

Kuratius commented:
Closing, as I am working on a version that doesn't require the HW accelerator, so this would end up redundant. I'll post updates in the issue on the main repo if there is anything interesting.
The integer sqrt()  implementation will likely be moved to its own repo.
